File: leetcode/202501/ZigzagConversion.go
package main

import "strings"

func main() {
	// 测试用例
	input := "PAYPALISHIRING"
	numRows := 3
	result := convert(input, numRows)
	println(result) // 输出: "PAHNAPLSIIGYIR"
}

func convert(s string, numRows int) string {
	
	if numRows == 1 || numRows >= len(s) {
		return s
	}

	zigzag := make([]string, numRows)
	row, step := 0, 1

	for _, i := range s {
		zigzag[row] += string(i)

		if row == 0 {
			step = 1
		} else if row == numRows - 1 {
			step = -1
		}
		row += step
	}
	// return strings.Join(zigzag, "")
	var res strings.Builder
	for i := 0; i < numRows; i++ {
		res.WriteString(zigzag[i])
	}
	return res.String()
}

File: leetcode/Daily/ZigzagConversion/ZigzagConversion.py
# -*- encoding: utf-8 -*-
"""
@Introduce  : 
@File       : ZigzagConversion.py
@Author     : ryrl
@Email      : ryrl970311@gmail.com
@Time       : 2025/01/10 12:43
@Description: 
"""

def convert(s: str, numRows: int) -> str:

    if numRows == 1 or numRows >= len(s):
        return s
    
    zigzag = ['' for _ in range(numRows)]  # generate numRows number of strings
    row, step = 0, 1  # row is the current row, step is the direction of the row

    for i in s:

        zigzag[row] += i

        if row == 0:  
            step = 1
        
        elif row == numRows - 1:  # if the row is the last row, change the direction
            step = -1
        
        row += step  # move to the next row
    
    return ''.join(zigzag)